contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1047 @@
+    }
+    Schedule storage schedule = _scheduleIdToScheduleStruct[projectId];
+    // slither-disable-next-line unused-return return value irrelevant
+    _allScheduleIds.add(projectId);
+    RestrictionDuration
+      memory restrictionDuration = getRestrictionDurationForMethodologyAndVersion(
+        scheduleData.methodology,
+        scheduleData.methodologyVersion
+      );
+    if (!restrictionDuration.wasSet) {
+      revert RestrictionDurationNotSet({projectId: projectId});
+    }
+    // slither-disable-next-line unused-return return value irrelevant
+    _addressToScheduleIdSet[recipient].add(projectId);
+    schedule.exists = true;
+    schedule.startTime = scheduleData.startTime;
+    schedule.endTime = scheduleData.startTime + restrictionDuration.duration;
+    schedule.releasedAmountFloor = 0;
+    emit ScheduleCreated(projectId, schedule.startTime, schedule.endTime);
+  }
+
+  /**
+   * @notice Revokes unreleased tokens from a schedule.
+   * This is an *admin* operation callable only by addresses having TOKEN_REVOKER_ROLE
+   * (enforced in `batchRevokeUnreleasedTokenAmounts`)
+   *
+   * @dev Only the maximum revocable number of tokens (unreleased tokens) can be revoked.
+   * Once the tokens have been revoked, the current released amount can never fall below
+   * its current level, even if the linear release schedule of the new amount would cause
+   * the released amount to be lowered at the current timestamp (a floor is established).
+   *
+   * Unlike in the `withdrawFromSchedule` function, here we burn `RestrictedNORI`
+   * from the schedule owner but send that `BridgedPolygonNORI` back to Nori's
+   * treasury or an address of Nori's choosing (the *to* address).
+   * The *claimedAmount* is not changed because this is not a claim operation.
+   */
+  function _revokeUnreleasedTokens(
+    address to,
+    uint256 removalId,
+    uint256 amount
+  ) internal {
+    // slither-disable-next-line calls-loop
+    uint256 scheduleId = _removal.getProjectIdForRemoval(removalId);
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    if (!schedule.exists) {
+      revert NonexistentSchedule({scheduleId: scheduleId});
+    }
+    uint256 quantityToRevoke;
+    // amount of zero indicates revocation of all remaining tokens.
+    if (amount > 0) {
+      if (!(amount <= revocableQuantityForSchedule(scheduleId))) {
+        revert InsufficientUnreleasedTokens({scheduleId: scheduleId});
+      }
+      quantityToRevoke = amount;
+    } else {
+      quantityToRevoke = revocableQuantityForSchedule(scheduleId);
+    }
+
+    // burn correct proportion from each token holder
+    EnumerableSetUpgradeable.AddressSet storage tokenHolders = schedule
+      .tokenHolders;
+    uint256[] memory quantitiesToBurnForHolders = new uint256[](
+      tokenHolders.length()
+    );
+
+    uint256 cumulativeQuantityToBurn = 0;
+    for (uint256 i = 0; i < (tokenHolders.length() - 1); i++) {
```

Comment:
why is this for loop `length - 1` but the subsequent is not?

Reply:
e.g., cant line 8666 just check if `i === length`? is this a gas efficiency trick?

Reply:
the reason this loop is like this has to do with math precision.  since we're taking a total quantity to revoke and we're splitting it out into the correct amount to revoke on a per-holder basis, there is some division happening in which safe math rounding errors might shave off a wei or so from any given token holder's quantity to revoke.  In order to guarantee that the sum of each individual revocation is EXACTLY the same as the total quantity to revoke, we calculate the last balance to revoke by subtracting the running total from the final total. i'll leave a comment to make this more clear